diff --git a/org/apache/jasper/EmbeddedServletOptions.java b/org/apache/jasper/EmbeddedServletOptions.java
index 17b75c2..67b52f0 100644
--- a/org/apache/jasper/EmbeddedServletOptions.java
+++ b/org/apache/jasper/EmbeddedServletOptions.java
@@ -200,6 +200,11 @@ public final class EmbeddedServletOptions implements Options {
     private String compilerSourceVM = "1.5";
 
     /**
+     * The compiler class name.
+     */
+    private String compilerClassName = null;
+
+    /**
      * Cache for the TLD locations
      */
     private TldLocationsCache tldLocationsCache = null;
@@ -410,6 +415,13 @@ public final class EmbeddedServletOptions implements Options {
         return compilerSourceVM;
     }
 
+    /**
+     * Java compiler class to use.
+     */
+    public String getCompilerClassName() {
+        return compilerClassName;
+    }
+
     public boolean getErrorOnUseBeanInvalidClassAttribute() {
         return errorOnUseBeanInvalidClassAttribute;
     }
@@ -598,6 +610,12 @@ public final class EmbeddedServletOptions implements Options {
             this.javaEncoding = javaEncoding;
         }
 
+        String compilerClassName =
+            config.getInitParameter("compilerClassName");
+        if (compilerClassName != null) {
+            this.compilerClassName = compilerClassName;
+        }
+
         String reloadIntervalString =
             config.getInitParameter("reload-interval");
         if (reloadIntervalString != null) {                   
diff --git a/org/apache/jasper/JspC.java b/org/apache/jasper/JspC.java
index b22419f..235a7d6 100644
--- a/org/apache/jasper/JspC.java
+++ b/org/apache/jasper/JspC.java
@@ -709,6 +709,13 @@ public class JspC implements Options {
         compilerSourceVM = vm;
     }
 
+    /**
+     * @see Options#getCompilerClassName().
+     */
+    public String getCompilerClassName() {
+        return null;
+    }
+
     public TldLocationsCache getTldLocationsCache() {
 	return tldLocationsCache;
     }
diff --git a/org/apache/jasper/Options.java b/org/apache/jasper/Options.java
index 6d6abdb..167bc49 100644
--- a/org/apache/jasper/Options.java
+++ b/org/apache/jasper/Options.java
@@ -169,6 +169,11 @@ public interface Options {
     public String getCompilerSourceVM();   
 
     /**
+     * Jasper Java compiler class to use.
+     */
+    public String getCompilerClassName();
+
+    /**
      * The cache for the location of the TLD's
      * for the various tag libraries 'exposed'
      * by the web application.
diff --git a/org/apache/jasper/compiler/Compiler.java b/org/apache/jasper/compiler/Compiler.java
index ee30fed..70e8824 100644
--- a/org/apache/jasper/compiler/Compiler.java
+++ b/org/apache/jasper/compiler/Compiler.java
@@ -707,6 +707,14 @@ public class Compiler {
      * The default is to use javac from ant.
      */
     private void initJavaCompiler() throws JasperException {
+        if (options.getCompilerClassName() != null) {
+            Class c = getClassFor(options.getCompilerClassName());
+            try {
+                javaCompiler = (JavaCompiler) c.newInstance();
+            } catch (Exception ex) {
+            }
+        }
+        if (javaCompiler == null) {
     	boolean disablejsr199 = Boolean.TRUE.toString().equals(
     			System.getProperty("org.apache.jasper.compiler.disablejsr199"));
         Double version = 
@@ -726,6 +734,7 @@ public class Compiler {
                 }
             }
         }
+        }
         if (javaCompiler == null) {
             Class c = getClassFor("org.apache.tools.ant.taskdefs.Javac");
             if (c != null) {
diff --git a/org/apache/jasper/compiler/Generator.java b/org/apache/jasper/compiler/Generator.java
index 8908872..f60cca3 100644
--- a/org/apache/jasper/compiler/Generator.java
+++ b/org/apache/jasper/compiler/Generator.java
@@ -406,7 +406,7 @@ class Generator {
      * (shared by servlet and tag handler preamble generation)
      */
     private void genPreambleStaticInitializers() throws JasperException {
-        out.printil("private static final JspFactory _jspxFactory = JspFactory.getDefaultFactory();");
+        out.printil("private final JspFactory _jspxFactory = com.liferay.servletjspcompiler.jsp.JspFactoryFactory.getJspFactory();");
         out.println();
 
         // Static data for getDependants()
diff --git a/org/apache/jasper/compiler/JspRuntimeContext.java b/org/apache/jasper/compiler/JspRuntimeContext.java
index 0662065..4f1e9a2 100644
--- a/org/apache/jasper/compiler/JspRuntimeContext.java
+++ b/org/apache/jasper/compiler/JspRuntimeContext.java
@@ -146,8 +146,6 @@ public final class JspRuntimeContext implements Runnable {
                           + ex.getMessage(), ex);
             }
         }
-
-        JspFactory.setDefaultFactory(factory);
     }
 
     // ----------------------------------------------------------- Constructors
diff --git a/org/apache/jasper/compiler/Jsr199JavaCompiler.java b/org/apache/jasper/compiler/Jsr199JavaCompiler.java
index 38e9cec..8109fd0 100644
--- a/org/apache/jasper/compiler/Jsr199JavaCompiler.java
+++ b/org/apache/jasper/compiler/Jsr199JavaCompiler.java
@@ -92,7 +92,7 @@ public class Jsr199JavaCompiler implements JavaCompiler {
         // a JSP compilation can produce multiple class files, we need to
         // keep track of them all.
 
-    private ArrayList<String> options = new ArrayList<String>();
+    protected ArrayList<String> options = new ArrayList<String>();
     private CharArrayWriter charArrayWriter;
     private JspCompilationContext ctxt;
     private String javaFileName;
@@ -308,7 +308,7 @@ public class Jsr199JavaCompiler implements JavaCompiler {
         return classFile;
     }
 
-    private JavaFileManager getJavaFileManager(JavaFileManager fm) {
+    protected JavaFileManager getJavaFileManager(JavaFileManager fm) {
 
         return new ForwardingJavaFileManager<JavaFileManager>(fm) {
 
diff --git a/org/apache/jasper/compiler/TldLocationsCache.java b/org/apache/jasper/compiler/TldLocationsCache.java
index 694c67d..55647ab 100644
--- a/org/apache/jasper/compiler/TldLocationsCache.java
+++ b/org/apache/jasper/compiler/TldLocationsCache.java
@@ -317,7 +317,6 @@ public class TldLocationsCache {
         HashMap tldUriToLocationMap = (HashMap) ctxt.getAttribute(
             Constants.JSP_TLD_URI_TO_LOCATION_MAP);
         if (tldUriToLocationMap != null) {
-            localTldsProcessed = true;
             mappings = tldUriToLocationMap;
         } else {
             mappings = new HashMap();
diff --git a/org/apache/jasper/runtime/PageContextImpl.java b/org/apache/jasper/runtime/PageContextImpl.java
index 7400602..6ffbded 100644
--- a/org/apache/jasper/runtime/PageContextImpl.java
+++ b/org/apache/jasper/runtime/PageContextImpl.java
@@ -721,7 +721,12 @@ public class PageContextImpl extends PageContext {
             celResolver.add(new ListELResolver());
             celResolver.add(new ArrayELResolver());
             celResolver.add(new BeanELResolver());
+            if (com.liferay.portal.kernel.util.ServerDetector.isResin()) {
+                celResolver.add(new com.liferay.servletjspcompiler.el.JSP_2_2_CompatibleElResolver());
+            }
+            else {
             celResolver.add(new ScopedAttributeELResolver());
+            }
             elResolver = celResolver;
         }
         return elResolver;
